include/material/mohr_coulomb.tcc

```diff
@@ -0,0 +1,576 @@
+    (*dp_dsigma) = (dp_depsilon * depsilon_dsigma) + (dp_drho * drho_dsigma) +
+                   (dp_dtheta * dtheta_dsigma);
+    dp_dj = dp_drho * sqrt(2.);
+  }
+
+  // compute softening part
+  double dphi_dpstrain = 0.;
+  double dc_dpstrain = 0.;
+  if (yield_type == 2 && epds > peak_epds_ && epds < crit_epds_) {
+    dphi_dpstrain = (residual_friction_angle_ - friction_angle_) * PI / 180. /
+                    (crit_epds_ - peak_epds_);
+    dc_dpstrain = (residual_cohesion_ - cohesion_) / (crit_epds_ - peak_epds_);
+  }
+  double epsilon = (1. / sqrt(3.)) * (stress(0) + stress(1) + stress(2));
+  double df_dphi = sqrt(3. / 2.) * rho *
+                       ((sin(phi_) * sin(theta + PI / 3.) /
+                         (sqrt(3.) * cos(phi_) * cos(phi_))) +
+                        (cos(theta + PI / 3.) / (3. * cos(phi_) * cos(phi_)))) +
+                   (epsilon / (sqrt(3.) * cos(phi_) * cos(phi_)));
+  double df_dc = -1.;
+  (*softening) =
+      (-1.) * ((df_dphi * dphi_dpstrain) + (df_dc * dc_dpstrain)) * dp_dj;
+}
+
+//! Compute stress
+template <unsigned Tdim>
+Eigen::Matrix<double, 6, 1> mpm::MohrCoulomb<Tdim>::compute_stress(
+    const Vector6d& stress, const Vector6d& dstrain,
+    const ParticleBase<Tdim>* ptr, std::map<std::string, double>* state_vars) {
+  // Friction and dilation in radians
+  const double phi_max = friction_angle_ * PI / 180.;
+  const double psi_max = dilation_angle_ * PI / 180.;
+  const double c_max = cohesion_;
+  const double phi_min = residual_friction_angle_ * PI / 180.;
+  const double psi_min = residual_dilation_angle_ * PI / 180.;
+  const double c_min = residual_cohesion_;
+
+  // Current MC parameters using a linear softening rule
+  // plastic deviatoric strain
+  Eigen::Matrix<double, 6, 1> plastic_deviatoric_strain;
+  plastic_deviatoric_strain(0) = (*state_vars).at("epds0");
+  plastic_deviatoric_strain(1) = (*state_vars).at("epds1");
+  plastic_deviatoric_strain(2) = (*state_vars).at("epds2");
+  plastic_deviatoric_strain(3) = (*state_vars).at("epds3");
+  plastic_deviatoric_strain(4) = (*state_vars).at("epds4");
+  plastic_deviatoric_strain(5) = (*state_vars).at("epds5");
+
+  const double epds =
+      (2. / 3.) *
+      sqrt(3. / 2. *
+               (plastic_deviatoric_strain(0) * plastic_deviatoric_strain(0) +
+                plastic_deviatoric_strain(1) * plastic_deviatoric_strain(1) +
+                plastic_deviatoric_strain(2) * plastic_deviatoric_strain(2)) +
+           3. * (plastic_deviatoric_strain(3) * plastic_deviatoric_strain(3) +
+                 plastic_deviatoric_strain(4) * plastic_deviatoric_strain(4) +
+                 plastic_deviatoric_strain(5) * plastic_deviatoric_strain(5)));
+  Vector6d trial_stress = stress + (this->de_ * dstrain);
+
+  double j2 = 0;
+  double j3 = 0;
+  double rho = 0;
+  double theta = 0;
+  this->compute_rho_theta(stress, &j2, &j3, &rho, &theta);
+
+  double epsilon = (1. / sqrt(3.)) * (stress(0) + stress(1) + stress(2));
```

Comment:
Why is this epsilon, confusing as it relates to stress?

Reply:
We will remove the `epsilon` from the `compute_stress` function, becauce we can compute it in other functions as they have `stress`.